module/zfs/dbuf.c

```diff
@@ -1462,6 +1464,7 @@ dbuf_handle_indirect_hole(dmu_buf_impl_t *db, dnode_t *dn, blkptr_t *dbbp)
 	uint32_t indbs = 1ULL << dn->dn_indblkshift;
 	int n_bps = indbs >> SPA_BLKPTRSHIFT;
 
+	ASSERT(MUTEX_HELD(&db->db_mtx));
```

Comment:
This is practically a buffer fill.  It does not need `db` pointer at all.  It could be refactored the same as above, if you wish.

Reply:
How?  Are you suggesting just providing the blkptr_t* pointer to this function?  Its parent would still need the mutex, then.

Reply:
I suggest parent first allocate a buffer, then pass it to this function to fill and only then calls `dbuf_set_data()`.